### ISSUE TITLE
Add Stripe :)

### ISSUE THREAD

**Issue by ashkan18** [artsy/stress#8]

**Comment by ashkan18** [comment]
There are some discussions around this still going, opening to have the discussions here.

**Comment by joeyAghion** [comment]
Thanks and sorry for causing this to be re-opened.

I think this system will _have_ to interact with Stripe to accomplish its basic work. That includes things like temporary authorizations, captures, and refunds (but _not_ things like credit card tokenization and back-account-linking). Since Stripe's API is a well-established interface and this system will depend on it closely, I'd rather _not_ delegate all of that to Gravity. This also saves Gravity from having to know about things like temporary authorizations or charge metadata, which are purely order-related concerns.

This _would_ mean that Artsy's platform would interact with Stripe from multiple systems, but that's not a problem in my opinion. It may actually be a _feature_, allowing different clients, API versions, or background job strategies to be applied as necessary for different business needs. One problematic overlap would be webhooks, with separate systems each needing their own webhooks and possibly needing to ignore calls that relate to other systems' actions.

**Comment by ashkan18** [comment]
The part of Stripe interaction i really want to have centralized is card tokenization and merchant account creation. As long as thats stored in one place and multiple systems don't create customers and merchant accounts i would be 👍.
I agree that the call to stripe might be duplicated from multiple systems but that would have happened even if we had a separate system for processing charges. The only model that may get duplicated in systems calling Stripe would be transaction model which basically stores Stripe's response everytime we call them for charge. https://github.com/artsy/lewitt-api/blob/master/app/models/invoice_transaction.rb and https://github.com/artsy/gravity/blob/master/app/models/domain/order_transaction.rb

My biggest argument for having a separate system for processing charges is, if we ever want to support different type of processors (lets say paypal) all of these systems will also need to be updated, but imagine if there was a system that would get a `charge_processor_id` and `amount` and handles processing, that `charge_proceesor_id` could be a stripe card or paypal account and from Exchange/Lewitt perspective they don't care, processing service will know how to handle that.

If we are thinking we are going to only support Stripe for near future im good with adding Stripe here.

**Comment by joeyAghion** [comment]
I totally agree about potentially abstracting the payment provider, I'm just pretty sure I wouldn't want Gravity to be where that happens.

Regarding some code duplication like processing Stripe's responses into our transaction data models, you're right. Where there is real duplication or potential for savings via reuse, maybe we should extract a small library.
